Update Environments (#17740)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/environments.md

monikerRange: '>= aspnetcore-2.1'
ms.author: riande
ms.custom: mvc
ms.date: 4/12/2020
uid: fundamentals/environments
---
# Use multiple environments in ASP.NET Core

::: moniker range=">= aspnetcore-3.0"

By [Rick Anderson](https://twitter.com/RickAndMSFT) and [Kirk Larkin](https://twitter.com/serpent5)

ASP.NET Core configures app behavior based on the runtime environment using an environment variable.

[View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/fundamentals/environments/3.1sample) ([how to download](xref:index#how-to-download-a-sample))

## Environments

> ASP.NET Core reads the environment variable `ASPNETCORE_ENVIRONMENT` at app startup and stores the value in [IWebHostEnvironment.EnvironmentName](xref:Microsoft.Extensions.Hosting.IHostEnvironment.EnvironmentName). `ASPNETCORE_ENVIRONMENT` can be set to any value, but three values are provided by the framework:

Reviewer: Things are a bit different here now. The value used for `EnvironmentName` comes from:

1. `DOTNET_ENVIRONMENT`, which is loaded first.
1. `ASPNETCORE_ENVIRONMENT`, but only if `ConfigureWebHostDefaults` is used. This value overrides `DOTNET_ENVIRONMENT`.

This paragraph also talks about `IWebHostEnvironment`, but the property is declared in `IHostEnvironment`, which `IWebHostEnvironment` implements/extends. This is important for those using the generic host _without_ the web additions, because they'd only care about `IHostEnvironment`.

It looks like both of these observations might be relevant throughout the topic.

> * <xref:Microsoft.Extensions.Hosting.Environments.Development> : The default in development.

Reviewer: This is the default because it's set in `launchSettings.json`, but otherwise it would be `Production`. The description here reads to me as a kind of circular reference. "Development is the default in development" raises the question of what it means to be "in development".

* <xref:Microsoft.Extensions.Hosting.Environments.Staging>

> * <xref:Microsoft.Extensions.Hosting.Environments.Production> : The default outside of development.

Reviewer: I think this should say something more along the lines of how this is the default if a value isn't specified using either of the environment variables.

[!code-cshtml[](environments/sample-snapshot/EnvironmentsSample/Pages/About.cshtml)]

The [About page](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/fundamentals/environments/3.1sample/EnvironmentsSample/Pages/About.cshtml) from the [sample code](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/fundamentals/environments/3.1sample) includes the preceding markup and displays the value of `IWebHostEnvironment.EnvironmentName`.

On Windows and macOS, environment variables and values aren't case-sensitive. Linux environment variables and values are **case-sensitive** by default.

### Create EnvironmentsSample

The [sample code](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/fundamentals/environments/3.1sample) used in this document is based on a Razor Pages project named *EnvironmentsSample*.

The following code creates and runs a web app named *EnvironmentsSample*:

```bash
dotnet new webapp -o EnvironmentsSample
cd EnvironmentsSample
dotnet run --verbosity normal
```

> The preceding code displays some of the following output:

Reviewer: ```suggestion
When the app runs, it displays the following output:
```

Author: Edited your suggestion, add ***some of*** because I edited out most of the output.

      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: c:\tmp\EnvironmentsSample
```

### Development and launchSettings.json

The development environment can enable features that shouldn't be exposed in production. For example, the ASP.NET Core templates enable the [Developer Exception Page](xref:fundamentals/error-handling#developer-exception-page) in the development environment.

The environment for local machine development can be set in the *Properties\launchSettings.json* file of the project. Environment values set in *launchSettings.json* override values set in the system environment.

The *launchSettings.json* file:

* Is only used on the local development machine.
* Is not deployed.
* contains profile settings.

The following JSON shows the *launchSettings.json* file for an ASP.NET Core web projected named *EnvironmentsSample* created with Visual Studio or `dotnet new`:

[!code-json[](environments/3.1sample/EnvironmentsSample/Properties/launchSettingsCopy.json)]

The preceding markup contains two profiles:

> * `IIS Express`: The default profile used when launching the app from Visual Studio. The `"commandName"` key has the value `"IISExpress"`, therefore, [IISExpress](/iis/extensions/introduction-to-iis-express/iis-express-overview) is the web server. You can set the launch profile to the project or any other profile included. For example, in the image below, selecting the project name launches the [Kestrel server](xref:fundamentals/servers/kestrel).

Reviewer: ```suggestion
* `IIS Express`: The default profile used when launching the app from Visual Studio. The `"commandName"` key has the value `"IISExpress"`, therefore, [IISExpress](/iis/extensions/introduction-to-iis-express/iis-express-overview) is the web server. You can set the launch profile to the project or any other profile included. For example, in the image below, selecting the project name launches the [Kestrel web server](xref:fundamentals/servers/kestrel).
```

  ![IIS Express launch on menu](environments/_static/iisx2.png)
* `EnvironmentsSample`: The profile name is the project name. This profile is used by default when launching the app with `dotnet run`.  The `"commandName"` key has the value `"Project"`, therefore, the [Kestrel web server](xref:fundamentals/servers/kestrel) is launched.

The value of `commandName` can specify the web server to launch. `commandName` can be any one of the following:

* `IISExpress` : Launches IIS Express.
* `IIS` : No web server launched. IIS is expected to be available.
* `Project` : Launches Kestrel.

The Visual Studio project properties **Debug** tab provides a GUI to edit the *launchSettings.json* file. Changes made to project profiles may not take effect until the web server is restarted. Kestrel must be restarted before it can detect changes made to its environment.

![Project Properties Setting Environment variables](environments/_static/project-properties-debug.png)

The following *launchSettings.json* file contains multiple profiles:

[!code-json[](environments/3.1sample/EnvironmentsSample/Properties/launchSettings.json)]

> Profiles can be selected from Visual Studio or from `dotnet run --launch-profile <Profile Name>`.

Reviewer: ```suggestion
Profiles can be selected from Visual Studio or using `dotnet run --launch-profile <Profile Name>`.
```

monikerRange: '>= aspnetcore-2.1'
ms.author: riande
ms.custom: mvc
ms.date: 4/12/2020
uid: fundamentals/environments
---
# Use multiple environments in ASP.NET Core

::: moniker range=">= aspnetcore-3.0"

By [Rick Anderson](https://twitter.com/RickAndMSFT) and [Kirk Larkin](https://twitter.com/serpent5)

ASP.NET Core configures app behavior based on the runtime environment using an environment variable.

[View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/fundamentals/environments/3.1sample) ([how to download](xref:index#how-to-download-a-sample))

## Environments

> ASP.NET Core reads the environment variable [IHostEnvironment.EnvironmentName](xref:Microsoft.Extensions.Hosting.IHostEnvironment.EnvironmentName) value on app startup from the environment variable:

Reviewer: ```suggestion
To determine the runtime environment, ASP.NET Core reads from the following environment variables:
```

1. [DOTNET_ENVIRONMENT](xref:fundamentals/configuration#default-host-configuration)
1. `ASPNETCORE_ENVIRONMENT` when <xref:Microsoft.Extensions.Hosting.GenericHostBuilderExtensions.ConfigureWebHostDefaults*> is called. The default ASP.NET Core web app templates call `ConfigureWebHostDefaults`. The `ASPNETCORE_ENVIRONMENT` value overrides `DOTNET_ENVIRONMENT`.

> `IHostEnvironment.EnvironmentName` can be set to any value, but three values are provided by the framework:

Reviewer: ```suggestion
`IHostEnvironment.EnvironmentName` can be set to any value, but the following values are provided by the framework:
```

> `IHostEnvironment.EnvironmentName` can be set to any value, but three values are provided by the framework:

Reviewer: ```suggestion
Use [IHostEnvironment.EnvironmentName](xref:Microsoft.Extensions.Hosting.IHostEnvironment.EnvironmentName) or `IWebHostEnvironment.EnvironmentName` to access the runtime environment name.

The runtime environment name can be set to any value, but the following values are provided by the framework:
```

> * <xref:Microsoft.Extensions.Hosting.Environments.Development> : The [launchSettings.json](#lsj) file sets `EnvironmentName` to `Development` on the local machine.

Reviewer: ```suggestion
* <xref:Microsoft.Extensions.Hosting.Environments.Development> : The [launchSettings.json](#lsj) file sets `ASPNETCORE_ENVIRONMENT` to `Development` on the local machine.
```

* <xref:Microsoft.Extensions.Hosting.Environments.Staging>

> * <xref:Microsoft.Extensions.Hosting.Environments.Production> : The default if `DOTNET_ENVIRONMENT` or `ASPNETCORE_ENVIRONMENT` has not been set.

Reviewer: ```suggestion
* <xref:Microsoft.Extensions.Hosting.Environments.Production> : The default if `DOTNET_ENVIRONMENT` and `ASPNETCORE_ENVIRONMENT` have not been set.
```

The following code:

* Calls [UseDeveloperExceptionPage](/dotnet/api/microsoft.aspnetcore.builder.developerexceptionpageextensions.usedeveloperexceptionpage) when `ASPNETCORE_ENVIRONMENT` is set to `Development`.
* Calls [UseExceptionHandler](/dotnet/api/microsoft.aspnetcore.builder.exceptionhandlerextensions.useexceptionhandler) when the value of `ASPNETCORE_ENVIRONMENT` is set to `Staging`, `Production`, or  `Staging_2`.
* Injects <xref:Microsoft.AspNetCore.Hosting.IWebHostEnvironment> into `Startup.Configure`. This approach is useful when the app only requires adjusting `Startup.Configure` for a few environments with minimal code differences per environment.
* Is similar to the code generated by the ASP.NET Core templates.

[!code-csharp[](environments/3.1sample/EnvironmentsSample/Startup.cs?name=snippet)]

> The [Environment Tag Helper](xref:mvc/views/tag-helpers/builtin-th/environment-tag-helper) uses the value of [IWebHostEnvironment.EnvironmentName](xref:Microsoft.Extensions.Hosting.IHostEnvironment.EnvironmentName) to include or exclude markup in the element:

Reviewer: ```suggestion
The [Environment Tag Helper](xref:mvc/views/tag-helpers/builtin-th/environment-tag-helper) uses the value of [IHostEnvironment.EnvironmentName](xref:Microsoft.Extensions.Hosting.IHostEnvironment.EnvironmentName) to include or exclude markup in the element:
```

1. Select **OK** to close the **Add/Edit application setting** window.
1. Select **Save** at the top of the **Configuration** blade.

Azure App Service automatically restarts the app after an app setting is added, changed, or deleted in the Azure portal.

### Windows

Environment values in *launchSettings.json* override values set in the system environment.

To set the `ASPNETCORE_ENVIRONMENT` for the current session when the app is started using [dotnet run](/dotnet/core/tools/dotnet-run), the following commands are used:

**Command prompt**

> ```console zz

Reviewer: ```suggestion
```console
```

### Configuration by environment

To load configuration by environment, we recommend:

* *appsettings* files (*appsettings.{Environment}.json*). See <xref:fundamentals/configuration/index#json-configuration-provider>.
* Environment variables (set on each system where the app is hosted). See <xref:fundamentals/host/generic-host#environmentname> and <xref:security/app-secrets#environment-variables>.
* Secret Manager (in the Development environment only). See <xref:security/app-secrets>.
To load configuration by environment, see <xref:fundamentals/configuration/index#json-configuration-provider>.

## Environment-based Startup class and methods

### Inject IWebHostEnvironment into Startup.Configure

Inject <xref:Microsoft.AspNetCore.Hosting.IWebHostEnvironment> into `Startup.Configure`. This approach is useful when the app only requires adjusting `Startup.Configure` for a few environments with minimal code differences per environment.

```csharp
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{
    if (env.IsDevelopment())
    {
        // Development environment code
    }
    else
    {
        // Code for all other environments
    }
}

> ```

Reviewer: I think it's still useful to show injecting `IWebHostEnvironment` into `Startup.Configure`. This is what the templates do and it's a bit easier if you only care about the environment when setting up the middleware pipeline.

Author: Added right after I show this in a snippet:

The preceding code:

* Injects <xref:Microsoft.AspNetCore.Hosting.IWebHostEnvironment> into `Startup.Configure`. This approach is useful when the app only requires adjusting `Startup.Configure` for a few environments with minimal code differences per environment.
* Is similar to the code generated by the ASP.NET Core templates.
